server/src/main/kotlin/core/gameflow/Dealer.kt

```diff
@@ -0,0 +1,59 @@
+package core.gameflow
+
+import core.cards.Card
+import core.cards.baseDeck
+import kotlin.random.Random
+
+class Dealer() {
+    private var deckIterator: ListIterator<Card>? = null
+
+    fun dealHoleCards(handState: HandState): HandState {
+        deckIterator!!
+
+        for (player: Player in handState.players)
+            assert(player.holeCards.isEmpty())
+
+        val updatedPlayers = handState.players.map { it.withCards(listOf(deckIterator!!.next(), deckIterator!!.next())) }
+        return handState.copy(players = updatedPlayers)
+    }
+
+    fun dealFlop(handState: HandState): HandState {
+        deckIterator!!
+
+        assert(handState.communityCards.isEmpty())
+
+        val flopCards = listOf(deckIterator!!.next(), deckIterator!!.next(), deckIterator!!.next())
+        return handState.copy(communityCards = flopCards)
+    }
+
+    fun dealTurn(handState: HandState): HandState {
+        deckIterator!!
+
+        assert(handState.communityCards.size == 3)
+
+        val turnCard = deckIterator!!.next()
+        return handState.copy(communityCards = handState.communityCards + turnCard)
+    }
+
+    fun dealRiver(handState: HandState): HandState {
+        deckIterator!!
+
+        assert(handState.communityCards.size == 4)
+
+        val riverCard = deckIterator!!.next()
+        return handState.copy(communityCards = handState.communityCards + riverCard)
+    }
+
+    fun shuffle(seed: Int? = null) {
+        deckIterator = when (seed) {
+            null -> baseDeck.shuffled().listIterator()
+            else -> baseDeck.shuffled(Random(seed)).listIterator()
+        }
+    }
+
+    fun setColdDeck(cards: List<Card>) {
+        assert(cards.size == 52)
+        assert(cards.distinct() == cards)
+        deckIterator = cards.listIterator()
+    }
+}
```

Comment:
I think it would be nicer if instead of checking the number of community cards, we would check if HandState.bettingRound is correct before dealing each street. Maybe it should be HandState's responsibility to guarantee that the community cards number and betting round are coherent.

Reply:
I agree, but where in handState should it be implemented?

Reply:
What about checking if the players have their holeCards field empty(in dealHoleCards method)?

Reply:
Maybe we shouldn't implement it at all. Let's not go super defensive with our code. I would just change assertions in dealer's methods to use HandState.bettingRound (for sake of code readability) and trust that the state that dealer operates on is valid (that's why we unit test other components of our app as well).

Reply:
Good point, I made necessary changes.